[SPARK-29519][SQL] SHOW TBLPROPERTIES should do multi-catalog resolution.

planga82 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Add ShowTablePropertiesStatement and make SHOW TBLPROPERTIES go through the same catalog/table resolution framework of v2 commands.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
It's important to make all the commands have the same table resolution behavior, to avoid confusing end-users. e.g.

USE my_catalog
DESC t // success and describe the table t from my_catalog
SHOW TBLPROPERTIES t // report table not found as there is no table t in the session catalog

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->
yes. When running SHOW TBLPROPERTIES Spark fails the command if the current catalog is set to a v2 catalog, or the table name specified a v2 catalog.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Unit tests.

AmplabJenkins commented:
Can one of the admins verify this patch?

planga82 commented:
cc: @cloud-fan @rdblue @viirya @imback82

AmplabJenkins commented:
Can one of the admins verify this patch?

viirya commented:
ok to test

SparkQA commented:
**[Test build #112320 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112320/testReport)** for PR 26176 at commit [`4bd0c5a`](https://github.com/apache/spark/commit/4bd0c5afa5476757ce40b196284f7f8876423d5a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17300/
Test PASSed.

SparkQA commented:
**[Test build #112320 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112320/testReport)** for PR 26176 at commit [`4bd0c5a`](https://github.com/apache/spark/commit/4bd0c5afa5476757ce40b196284f7f8876423d5a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowTablePropertiesStatement(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112320/
Test FAILed.

SparkQA commented:
**[Test build #112335 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112335/testReport)** for PR 26176 at commit [`4bc65b1`](https://github.com/apache/spark/commit/4bc65b1427d1384837d8dde091b26a0fe6d2a72c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17314/
Test PASSed.

SparkQA commented:
**[Test build #112335 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112335/testReport)** for PR 26176 at commit [`4bc65b1`](https://github.com/apache/spark/commit/4bc65b1427d1384837d8dde091b26a0fe6d2a72c).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112335/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17317/
Test PASSed.

SparkQA commented:
**[Test build #112338 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112338/testReport)** for PR 26176 at commit [`f651227`](https://github.com/apache/spark/commit/f651227653c68a214311c59c0af68600170e3956).

SparkQA commented:
**[Test build #112338 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112338/testReport)** for PR 26176 at commit [`f651227`](https://github.com/apache/spark/commit/f651227653c68a214311c59c0af68600170e3956).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112338/
Test PASSed.

imback82 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2728,4 +2728,20 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+  override def visitShowTblProperties(
+      ctx: ShowTblPropertiesContext): LogicalPlan = withOrigin(ctx) {
+    ShowTablePropertiesStatement(
+      visitMultipartIdentifier(ctx.multipartIdentifier()),
```
`ctx.table` since the alias is available? I am not sure what the preferred way is, and for commands like `ALTER TABLE`, we haven't used the alias `table` in `SqlBase.g4`.  @cloud-fan can confirm.

planga82 replied:
I have seen the other pull requests as example. If we have to change it, it's not a problem.

cloud-fan replied:
It's not a big deal. We can either remove the `table` alias, or use `ctx.table` here.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:
```diff
@@ -1210,6 +1210,23 @@ class DataSourceV2SQLSuite
+      val e1 = intercept[AnalysisException] {
+        sql(s"SHOW TBLPROPERTIES $t")
+      }
+      assert(e1.message.contains("SHOW TBLPROPERTIES is only supported with v1 tables"))
```
I think we can implement SHOW TBLPROPERTIES with v2 APIs. You can refer to https://github.com/apache/spark/pull/26166 about how to implement a v2 command.

planga82 commented:
V2 command implemented

SparkQA commented:
**[Test build #113301 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113301/testReport)** for PR 26176 at commit [`fb93e0f`](https://github.com/apache/spark/commit/fb93e0f740eca41e688571c9c2ff05586334cbbd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18207/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:
```diff
@@ -371,3 +371,15 @@ case class ShowCurrentNamespace(catalogManager: CatalogManager) extends Command
+/**
+ * The logical plan of the SHOW TBLPROPERTIES command that works for v2 catalogs.
+ */
+case class ShowTblproperties(
```
let's use the full name: `ShowTableProperties`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCatalogs.scala:
```diff
@@ -207,6 +207,10 @@ class ResolveCatalogs(val catalogManager: CatalogManager)
       ShowCurrentNamespace(catalogManager)
+
+    case ShowTablePropertiesStatement(NonSessionCatalog(catalog, tableName), propertyKey) =>
+      ShowTblproperties(catalog.asTableCatalog, tableName.asIdentifier, propertyKey)
```
The logical plan should take a `NamedRelation`, and here we just pass in a `UnresolvedV2Relation`.

You can take a look at how `DescribeTableStatement` is handled in this file.

planga82 replied:
Ok, I see, the Analizer resolve tables in V2 catalog, thanks!!

SparkQA commented:
**[Test build #113301 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113301/testReport)** for PR 26176 at commit [`fb93e0f`](https://github.com/apache/spark/commit/fb93e0f740eca41e688571c9c2ff05586334cbbd).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowTablePropertiesStatement(`
  * `case class ShowTblproperties(`
  * `case class ShowTblpropertiesExec(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113301/
Test FAILed.

SparkQA commented:
**[Test build #113400 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113400/testReport)** for PR 26176 at commit [`706a555`](https://github.com/apache/spark/commit/706a555e968382837eb42b05f25f43c25ff288a0).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18298/
Test PASSed.

SparkQA commented:
**[Test build #113400 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113400/testReport)** for PR 26176 at commit [`706a555`](https://github.com/apache/spark/commit/706a555e968382837eb42b05f25f43c25ff288a0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowTableProperties(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113400/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCatalogs.scala:
```diff
@@ -207,6 +207,12 @@ class ResolveCatalogs(val catalogManager: CatalogManager)
       ShowCurrentNamespace(catalogManager)
+
+    case ShowTablePropertiesStatement(
+    nameParts @ NonSessionCatalog(catalog, tableName), propertyKey) =>
```
nit: 4 space indentation before `nameParts @`

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:
```diff
@@ -1452,6 +1452,49 @@ class DataSourceV2SQLSuite
+      spark.sql(s"CREATE TABLE $t (id bigint, data string) USING $provider " +
+        s"TBLPROPERTIES ('owner'='$owner', 'status'='$status')")
+
+      val tbl1 = sql(s"SHOW TBLPROPERTIES $t ('status')")
```
can we also test if the key not found?

planga82 replied:
Ok

SparkQA commented:
**[Test build #113466 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113466/testReport)** for PR 26176 at commit [`93ec3f8`](https://github.com/apache/spark/commit/93ec3f826f83d9071386be198147fc320f9a583a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18358/
Test PASSed.

imback82 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCatalogs.scala:
```diff
@@ -207,6 +207,12 @@ class ResolveCatalogs(val catalogManager: CatalogManager)
+      nameParts @ NonSessionCatalog(catalog, tableName), propertyKey) =>
+        val r = UnresolvedV2Relation(nameParts, catalog.asTableCatalog, tableName.asIdentifier)
+        ShowTableProperties(r, propertyKey)
+
```
nit: remove extra empty line.

imback82 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -3067,4 +3067,20 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+   * corresponding values are returned.
+   * The syntax of using this command in SQL is:
+   * {{{
+   *   SHOW TBLPROPERTIES table_name[('propertyKey')];
```
nit: `table_name` -> `multi_part_name`

imback82 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowTblpropertiesExec.scala:
```diff
@@ -0,0 +1,51 @@
+          encoder.toRow(new GenericRowWithSchema(Array(k, properties(k)), schema)).copy()).toSeq
+    }
+  }
+
```
nit: remove empty line

imback82 commented on sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:
```diff
@@ -1452,6 +1452,64 @@ class DataSourceV2SQLSuite
+
+      assert(tbl1.schema === schema)
+      assert(expected === tbl1.collect())
+
```
nit: remove empty line

imback82 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowTblpropertiesExec.scala:
```diff
@@ -0,0 +1,51 @@
+/**
+ * Physical plan node for showing tblproperties.
+ */
+case class ShowTblpropertiesExec(
```
`ShowTablePropertiesExec` to match with the statement, etc. Same applies to other places below.

imback82 commented on sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:
```diff
@@ -1452,6 +1452,64 @@ class DataSourceV2SQLSuite
+      spark.sql(s"CREATE TABLE $t (id bigint, data string) USING $provider " +
+        s"TBLPROPERTIES ('owner'='$owner', 'status'='$status')")
+
+      val tbl1 = sql(s"SHOW TBLPROPERTIES $t")
```
Why `1`? How about just `properties`?

planga82 replied:
Changes done

SparkQA commented:
**[Test build #113466 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113466/testReport)** for PR 26176 at commit [`93ec3f8`](https://github.com/apache/spark/commit/93ec3f826f83d9071386be198147fc320f9a583a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113466/
Test PASSed.

SparkQA commented:
**[Test build #113471 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113471/testReport)** for PR 26176 at commit [`e5e44c1`](https://github.com/apache/spark/commit/e5e44c122f7aa27a7340f0a41bd9376f52e4aa33).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18363/
Test PASSed.

SparkQA commented:
**[Test build #113471 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113471/testReport)** for PR 26176 at commit [`e5e44c1`](https://github.com/apache/spark/commit/e5e44c122f7aa27a7340f0a41bd9376f52e4aa33).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113471/
Test PASSed.

imback82 reviewed (approved):
LGTM. Thanks @planga82!

planga82 commented:
Thanks for your comments!!

planga82 committed:
[SPARK-29519] SHOW TBLPROPERTIES should do multi-catalog resolution.


[SPARK-29519] Fix tests


[SPARK-29519] Resolve catalog in Analizer


[SPARK-29519] Fix pull request comments


[SPARK-29519] Fix pull request comments

planga82 commented:
Conflict resolved

SparkQA commented:
**[Test build #113547 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113547/testReport)** for PR 26176 at commit [`863ed90`](https://github.com/apache/spark/commit/863ed909832a93d2c65e80070b8a992dcc9e96e7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18436/
Test PASSed.

SparkQA commented:
**[Test build #113547 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113547/testReport)** for PR 26176 at commit [`863ed90`](https://github.com/apache/spark/commit/863ed909832a93d2c65e80070b8a992dcc9e96e7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowTablePropertiesStatement(`
  * `case class ShowTableProperties(`
  * `case class ShowTablePropertiesExec(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113547/
Test PASSed.

fuwhu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -691,6 +691,11 @@ class Analyzer(
+        CatalogV2Util.loadRelation(u.catalog, u.tableName)
+          .map(rel => show.copy(table = rel))
+          .getOrElse(u)
+
```
Why is it '.getOrElse(u)' instead of '.getOrElse(show)' here ?

cloud-fan replied:
good catch!

planga82 replied:
oh, yes good catch!!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCatalogs.scala:
```diff
@@ -207,6 +207,11 @@ class ResolveCatalogs(val catalogManager: CatalogManager)
       ShowCurrentNamespace(catalogManager)
+
+    case ShowTablePropertiesStatement(
+      nameParts @ NonSessionCatalog(catalog, tableName), propertyKey) =>
```
nit: 4 space indentation

planga82 committed:
[SPARK-29519] Fix analizer

SparkQA commented:
**[Test build #113599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113599/testReport)** for PR 26176 at commit [`846e21a`](https://github.com/apache/spark/commit/846e21a96e106f44fa0b88ab71d07095995b3790).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18485/
Test PASSed.

SparkQA commented:
**[Test build #113599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113599/testReport)** for PR 26176 at commit [`846e21a`](https://github.com/apache/spark/commit/846e21a96e106f44fa0b88ab71d07095995b3790).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113599/
Test PASSed.

cloud-fan commented:
thanks, merging to master!
